projects/python-client/src/datapane/client/api/ipython_utils.py

```diff
@@ -19,10 +25,219 @@ def block_to_iframe(block: BaseElement) -> str:
+        notebook_json = get_colab_notebook_json()
+    elif "VSCODE_PID" in os.environ:
+        notebook_json = get_vscode_notebook_json()
+    else:
+        notebook_json = get_jupyter_notebook_json()
+
+    return notebook_json
+
+
+def markdown_cell_to_block(cell: dict) -> Text:
+    """Convert a IPython notebook cell to a Datapane Text Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import Text
+
+    block = Text("".join(cell["source"]))
+
+    return block
+
+
+def input_cell_to_block(cell: dict) -> Code:
+    """Convert a IPython notebook cell to a Datapane Code Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import Code
+
+    block = Code("".join(cell["source"]))
+
+    return block
+
+
+def output_cell_to_block(cell: dict, ipython_output_cache: dict) -> typing.Any[BaseElement, None]:
+    """Convert a IPython notebook output cell to a Datapane Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import wrap_block
+
+    # Get the output object from the IPython output cache
+    cell_output_object = ipython_output_cache.get(cell["execution_count"], None)
+
+    # If there's corresponding output object, skip
+    if cell_output_object is None:
+        return None
+
+    try:
+        block = wrap_block(cell_output_object)
+        return block
+    except Exception:
+        return None
+
+
+@capture_event("IPython Cells to Blocks")
+def cells_to_blocks(ipython_output_cache: dict, opt_out: bool = True) -> typing.List[BaseElement]:
+    """Convert IPython notebook cells to a list of Datapane Blocks
+
+    Recognized cell tags:
+        - `dp-exclude` - Exclude this cell (when opt_out=True)
+        - `dp-include` - Include this cell (when opt_out=False)
+        - `dp-show-code` - Show the input code for this cell
+
+    Args:
+        ipython_output_cache: The output cache (Out or _oh) dict from a IPython notebook
+        opt_out: When True, all cells are converted to blocks unless explicitly opted out with the tag `dp-exclude`. When False, only cells with the tag `dp-include` are converted to blocks.
+
+    Returns:
+        List of Datapane Blocks
+
+    ..note:: IPython output caching must be enabled for this function to work. It is enabled by default.
+    """
+    display_msg("Converting cells to blocks.")
+
+    notebook_json = get_notebook_json()
+
+    blocks = []
+
+    for cell in notebook_json["cells"]:
+        tags = cell["metadata"].get("tags", [])
+
+        if (opt_out and "dp-exclude" not in tags) or (not opt_out and "dp-include" in tags):
+            block = None  # type: typing.Any[BaseElement, None]
+
+            if cell["cell_type"] == "markdown":
+                block = markdown_cell_to_block(cell)
+                blocks.append(block)
+
```

Comment:
empty line confusing in middle of `if/else` statement - I assumed that `black` would have changed that